consensus/src/twins/twins_node.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters,
+    epoch_manager::EpochManager,
+    network::NetworkTask,
+    network_interface::{ConsensusNetworkEvents, ConsensusNetworkSender},
+    network_tests::{NetworkPlayground, TwinId},
+    test_utils::{MockStateComputer, MockStorage, MockTransactionManager},
+    util::time_service::ClockTimeService,
+};
+use channel::{self, libra_channel, message_queues::QueueStyle};
+use consensus_types::common::{Author, Payload, Round};
+use futures::channel::mpsc;
+use libra_config::{
+    config::{
+        ConsensusProposerType::{self, RoundProposer},
+        NodeConfig, WaypointConfig,
+    },
+    generator::{self, ValidatorSwarm},
+};
+use libra_mempool::mocks::MockSharedMempool;
+use libra_types::{
+    ledger_info::LedgerInfoWithSignatures,
+    on_chain_config::{OnChainConfig, OnChainConfigPayload, ValidatorSet},
+    validator_info::ValidatorInfo,
+    waypoint::Waypoint,
+};
+use network::{
+    peer_manager::{conn_notifs_channel, ConnectionRequestSender, PeerManagerRequestSender},
+    protocols::network::{NewNetworkEvents, NewNetworkSender},
+};
+use std::{collections::HashMap, num::NonZeroUsize, sync::Arc};
+use tokio::runtime::{Builder, Runtime};
+
+/// Auxiliary struct that is preparing SMR for the test
+pub struct SMRNode {
+    pub id: TwinId,
+    pub storage: Arc<MockStorage>,
+    pub commit_cb_receiver: mpsc::UnboundedReceiver<LedgerInfoWithSignatures>,
+    _runtime: Runtime,
+    _shared_mempool: MockSharedMempool,
+    _state_sync: mpsc::UnboundedReceiver<Payload>,
+}
+
+fn author_from_config(config: &NodeConfig) -> Author {
+    config.validator_network.as_ref().unwrap().peer_id()
+}
+
+impl SMRNode {
+    fn start(
+        playground: &mut NetworkPlayground,
+        config: NodeConfig,
+        storage: Arc<MockStorage>,
+        twin_id: TwinId,
+    ) -> Self {
+        let (network_reqs_tx, network_reqs_rx) =
+            libra_channel::new(QueueStyle::FIFO, NonZeroUsize::new(8).unwrap(), None);
+        let (connection_reqs_tx, _) =
+            libra_channel::new(QueueStyle::FIFO, NonZeroUsize::new(8).unwrap(), None);
+        let (consensus_tx, consensus_rx) =
+            libra_channel::new(QueueStyle::FIFO, NonZeroUsize::new(8).unwrap(), None);
+        let (_conn_mgr_reqs_tx, conn_mgr_reqs_rx) = channel::new_test(8);
+        let (_, conn_notifs_channel) = conn_notifs_channel::new();
+        let network_sender = ConsensusNetworkSender::new(
+            PeerManagerRequestSender::new(network_reqs_tx),
+            ConnectionRequestSender::new(connection_reqs_tx),
+        );
+        let network_events = ConsensusNetworkEvents::new(consensus_rx, conn_notifs_channel);
+
```

Comment:
oh man... channels? for a single threaded framework? Is there no way to make this less complex :P

have you thought about making twins a wrapper around RoundProcessor (or whatever it is called these days?).

Reply:
you know networks are all about channels lol

Reply:
Ideally we shouldn't be using networks for this type of test unless you intend on running this on a real network